Experimental API #1 - tiny-atom/fusion

KidkArolis opened the pull request:
- action packs
- where each pack is of shape { state, actions }
- all gets/splits inside actions are scoped to that pack
- actions are bound to split and can be called as functions
- exposes a new method atom.fuse(actions) for extending the action packs

This is built purely on top of the tiny-atom/core with no modifications.
It demos a very decent default approach to working with tiny-atom. Whilst
keeping tiny-atom/core available for any alternative strategies.

Most of all, this preserves the combined effect/reducer approach of
tiny-atom, it preserves the custom merge strategy and is compatible
with all of the existing batteries log/devtool/react/preact.

Finally, it's backwards compatible, you should be able to utilise this in
an existing project gradually if no custom evolve was used.

KidkArolis committed:
Add experimental new api at tiny-atom/future

- action packs
- where each pack is of shape { state, actions }
- all gets/splits inside actions are scoped to that pack
- actions are bound to
- exposes a new method atom.fuse(actions) for extending the action packs

This is built purely on top of the tiny-atom/core with no modifications.
It demos a very decent default approach to working with tiny-atom. Whilst
keeping tiny-atom/core available for any alternative strategies.

Most of all, this preserves the combined effect/reducer approach of
tiny-atom, it preserves the custom merge strategy and is compatible
with all of the existing batteries log/devtool/react/preact.

Finally, it's backwards compatible, you should be able to utilise this in
an existing project gradually if no custom evolve was used.

KidkArolis commented:
See `examples/future-example` for a full TODO app demo.

But here's the gist of the idea:

```js
module.exports.todo = {
  state: {
    items: [
      'learn tiny-atom',
      'use tiny-atom',
      'star tiny-atom'
    ],
    input: ''
  },
  actions: {
    add: (get, split, item) => {
      const items = get().items.concat([item])
      split({ items, input: '' })
      split.analytics.track({ type: 'added' })
    },
    remove: (get, split, index) => {
      const items = get().items.filter((item, i) => i !== index)
      split({ items })
      split.analytics.track({ type: 'removed' })
    }
  }
}

module.exports.analytics = {
  state: {
    events: []
  },
  actions: {
    track: (get, split, event) => {
      const events = get().events.concat([event])
      split({ events })
    }
  }
}
```

* splits can be used as functions, e.g. `split.analytics.track`
* each action pack is namespaced to it's slice of the state by default
* it's possible to escape that via get by `get('analytics')` or `get({ root: true })`
* it's possible to extend the actions dynamically with `atom.fuse({ todo: { state, actions } }`

KidkArolis commented:
I'm contemplating whether we should make this the main api:

```js
const createAtom = require('tiny-atom')
```

And move the existing lower level api into:

```js
const createAtom = require('tiny-atom/core')
```

KidkArolis committed:
Add release-beta task

KidkArolis committed:
beta@1.2.0

KidkArolis commented:
You can try this with `yarn add tiny-atom@beta`

KwanMan commented:
Im really liking the `split()` and `split.onboarding()` way of calling actions on another part of the state. Why aren’t we doing that for the `get` as well to keep it consistent, `get()`/`get.analytics()` rather than `get()`/`get(‘analytics’)`

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

KwanMan commented:
Also not a fan of potentially effectively having 2 “cores”, anyone who stumbles onto this and wants to quickly use it has to effectively choose a core before even writing any code

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

KidkArolis committed:
Rename tiny-atom/future to tiny-atom/orb, improve get api

KidkArolis commented:
@KwanMan great suggestion! It's now:

```
get() // scoped to the action pack
get.todo()
get.hints()
```

Which is symmetric with:

```
split({ }) // scoped update
split.todo.add()
split.hints.hide()
```

KidkArolis commented:
I've renamed this from `future` to `orb`.

Tiny atom will remain what it is, which is a tiny flexible open state management solution at:

```js
require('tiny-atom')
```

And if you'd benefit from the namespaced locally scoped get/split composed from multiple action packs, good for your typical UI app - you can also just use the provided tiny-atom wrapper - `orb`:

```js
require('tiny-atom/orb')
```

In other words, your plain tiny-atom would be:

```js
const createAtom = require('tiny-atom')

const atom = createAtom(initialState, evolve, options)
```

But your typical usage might end up being:

```js
const createAtom = require('tiny-atom/orb')
const merge = require('tiny-atom/deep-merge')
const debug = require('tiny-atom/log')

const packs = {
  entities: { state, actions },
  feeds: { state, actions },
  picks: { state, actions }
}
const atom = createAtom(packs, { merge, debug })
```

This latter form - the orb - is basically a good recommended way to use tiny-atom, it's an opinionated approach - but it's _no longer flexible_. And therefore the core library `require('tiny-atom')`, which _is flexible_, is still the main package entry. It's all shipped in the same package, because `tiny-atom` is batteries included - you install it, and it will cater all your state management needs (structured orb, open core, logger, devtools, preact, react, merge helper, maybe more in the future).

And as a bonus - `orb` is created in such a way that you can switch to it from the basic version of tiny-atom in a backwards compatible way.

KwanMan commented on src/orb.js:
```diff
@@ -0,0 +1,129 @@
+  if (options.merge) {
+    return Object.keys(actions).reduce((state, namespace) => {
+      const pack = actions[namespace]
+      return typeof pack === 'function' ? state : options.merge(state, { [namespace]: pack.state })
```
why would this be a function?

KidkArolis replied:
This is to support the existing projects that already have large action objects:

```
const actions = {
  existingAction1: () =>{},
  existingAction2: () =>{},
  existingAction3: () =>{},

  onboarding: {
    state: {},
    action1: () => {},
    action2: () => {}
  }
}
```

KwanMan replied:
so you mean i can call `split.existingAction1()`? if so, which part of the state will that action be working on?

KidkArolis committed:
Rename tiny-atom/orb to tiny-atom/fusion

KidkArolis commented:
Renamed to `tiny-atom/fusion`. Since it's about fusing multiple action packs together. Reads better than `orb`.

KidkArolis commented:
Global - same as it is today.

This whole legacy mode can be turned off with options.strict flag. I think
ill turn strict mode on by default.


On Fri, 9 Mar 2018 at 13:35, Tommy Kwan <notifications@github.com> wrote:

> *@KwanMan* commented on this pull request.
> ------------------------------
>
> In src/orb.js
> <https://github.com/QubitProducts/tiny-atom/pull/32#discussion_r173450345>
> :
>
> > +
> +  const cache = {}
> +  const actions = {}
> +  const initialState = createInitialState(state, initialActions, options)
> +  const evolve = createEvolve(actions, cache, options)
> +  const atom = createAtomCore(initialState, evolve, options)
> +  atom.fuse = fuse(atom, actions, cache, options)
> +  atom.fuse(initialActions)
> +  return atom
> +}
> +
> +function createInitialState (state, actions, options) {
> +  if (options.merge) {
> +    return Object.keys(actions).reduce((state, namespace) => {
> +      const pack = actions[namespace]
> +      return typeof pack === 'function' ? state : options.merge(state, { [namespace]: pack.state })
>
> so you mean i can call split.existingAction1()? if so, which part of the
> state will that action be working on?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/QubitProducts/tiny-atom/pull/32#discussion_r173450345>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AATzWMy4MX924tUcw7MJrJuHR320v4Ckks5tcoU8gaJpZM4Sil2D>
> .
>

KidkArolis commented:
Will merge for history via #33. Closing this one.
